### ISSUE TITLE
wmlxgettext: unnecessary reordering of `#.` lines

### ISSUE THREAD

**Issue by stevecotton** [wesnoth/wesnoth#5989]
Each pot-update has many changes that are simply `#.` comment lines being reordered, it would be better if the tools had a stable sort of these lines. In contrast, the `#:` lines do seem to have a stable sort.

An example from the pre-1.15.15 pot update (df6e679f136142db1e012cff1aca64294b41086a):
```
@@ -4491,10 +4491,10 @@ msgstr ""
 msgid "Enemies"
 msgstr ""
 
-#. [side]: id=King Thurongar, type=Dwarvish Lord
-#. [side]: id=Melusand, type=Mermaid Diviner
 #. [side]
+#. [side]: id=King Thurongar, type=Dwarvish Lord
 #. [side]: id=Great Leader Darmog, type=Great Troll
+#. [side]: id=Melusand, type=Mermaid Diviner
 #: data/campaigns/Under_the_Burning_Suns/scenarios/06b_In_the_Domain_of_Dwarves.cfg:73
 #: data/campaigns/Under_the_Burning_Suns/scenarios/07a_Dealing_with_Dwarves.cfg:52
 #: data/campaigns/Under_the_Burning_Suns/scenarios/07b_Talking_with_Trolls.cfg:53
```

**Comment by cooljeanius** [comment]
oh yeah I've noticed this with my add-ons, too

**Comment by AncientLich** [comment]
I can't completely answer about pot-update, since it involves not only wmlxgettext tool, but also the gettext tool "msgmerge"  so the first thing to understand is if the issue is something related to wmlxgettext itself or not. If I remember well this kind of reorder is caused by msgmerge that sometimes shuffle them but I am not completely sure about it.
I am pretty confident, however, that wmlxgettext produces a specific order of comments both for #: and for #. comments. Infact Wmlxgettext uses a line-by-line parsing approach for each file in parsing list, following the order of files specified by user.
In pot-update target, in particular, wmlxgettext is launched to generate a blank pot update giving an explicit list of files to wmlxgettext itself and wmlxgettext will follow that list. Wmlxgettext than will start to parse file 1 and collect comments line-by-line following the order of lines in the source of file 1, then process file 2 in the same way and so on.
So wmlxgettext should always create exactly the same blank pot files if you invoke it in the same way parsing the same files in the same order, no matter how many times you will use it. Obliouvsly if you insert some line in the middle of code, or change the order of the parameters in a source file, wmlxgettext will change the order of the informations, because it will reproduce the order of the informations it founds on source files.
So we should understand if wmlxgettext actually creates the same blank pot file or not. If the blank pot file is the same, but the final .po file created with pot-update not, then the issue is somehow unfixable.... I mean is something that is not managable by wmlxgettext itself.
This because pot-update, when running msgmerge, will not run wmlxgettext anymore, and wmlxgettext has no control (and can't have any control) about what msgmerge do in order to generate the final po file.
That case, in order to "solve" the issue, it could be necessary to introduce a brand-new tool, separated from wmlxgettext, to be executed as a new third step AFTER merging po (after msgmerge)... but this thing would be very very trivial and hard to design and personally I would have no idea about how to develop a similar tool.

**Comment by stevecotton** [comment]
In a similar vein, another source of noise in pot-updates are the line numbers. The .po format and gettext's `--add-location` option support putting in just the source filenames without line numbers, it's just the gettext defaults to `--add-location=file:line`.

If no-one objects, then I'll look at removing those numbers.

**Comment by cooljeanius** [comment]
> In a similar vein, another source of noise in pot-updates are the line numbers. The .po format and gettext's `--add-location` option supports putting in just the source filenames without line numbers, it's just that gettext defaults to `--add-location=full`.
> 
> If no-one objects, then I'll look at removing those numbers.
> 
> Edit: it turned out to have a lot of (possibly one-off) noise: without the line numbers, the sort order changed. Although still grouped by file, the lines are then sorted by the msgid.

Yes the line numbers are noisy, but I imagine they might be helpful for translators trying to figure out the context for certain strings? I dunno, maybe wait and see if some translators have any input...

**Comment by stevecotton** [comment]
> I am pretty confident, however, that wmlxgettext produces a specific order of comments both for #: and for #. comments. Infact Wmlxgettext uses a line-by-line parsing approach for each file in parsing list, following the order of files specified by user.
In pot-update target, in particular, wmlxgettext is launched to generate a blank pot update giving an explicit list of files to wmlxgettext itself and wmlxgettext will follow that list.

I think that's correct, but that the list of files given to wmlxgettext is itself in a non-deterministic order. It's a one-liner to make wmlxgettext unconditionally sort the list, or a few more lines to add a `--sort-by-file` option. I'm planning to open a PR to make it sort unconditionally, but need to explain some odd stuff from testing first (see #6320 for an example).

**Comment by stevecotton** [comment]
Linking up the history, as some of this was done without PRs:

* Implemented in #6322.
* The merged PR was backported to 1.16 in 2adcb92fab.
* An option to use the order on the command line was added in #6586.
* That was forward-ported in 5c07604c016ac2e59a6f616033102270d339043a.
